pkg/controller/elasticsearch/nodespec/lifecycle_hook.go

```diff
@@ -146,62 +146,69 @@ function supports_node_shutdown() {
 version=""
 if [[ -f "{{.LabelsFile}}" ]]; then
   # get Elasticsearch version from the downward API
-  version=$(grep "{{.VersionLabelName}}" {{.LabelsFile}} | cut -d '=' -f 2)
+  version=$(grep "{{.VersionLabelName}}" "{{.LabelsFile}}" | cut -d '=' -f 2)
   # remove quotes
   version=$(echo "${version}" | tr -d '"')
 fi
 
 # if ES version does not support node shutdown exit early
 if ! supports_node_shutdown "$version"; then
-  delayed_exit 
+  delayed_exit
 fi
 
 # setup basic auth if credentials are available
 if [ -f "{{.PreStopUserPasswordPath}}" ]; then
-  PROBE_PASSWORD=$(<{{.PreStopUserPasswordPath}})
-  BASIC_AUTH="-u {{.PreStopUserName}}:${PROBE_PASSWORD}"
+  PROBE_PASSWORD=$(<"{{.PreStopUserPasswordPath}}")
+  BASIC_AUTH=("-u" "{{.PreStopUserName}}:${PROBE_PASSWORD}")
 else
   # typically the case on upgrades from versions that did not have this script yet and the necessary volume mounts are missing
   log "no API credentials available, will not attempt node shutdown orchestration from pre-stop hook"
   delayed_exit
 fi
 
-ES_URL={{.ServiceURL}}
+ES_URL="{{.ServiceURL}}"
 
 log "retrieving node ID"
-retry 10 request -X GET "$ES_URL/_cat/nodes?full_id=true&h=id,name" $BASIC_AUTH
-if [ "$?" -ne 0 ]; then
-	error_exit "failed to retrieve node ID"
+if ! retry 8 request -X GET "${ES_URL}/_cat/nodes?full_id=true&h=id,name" "${BASIC_AUTH[@]}"
```

Comment:
`8`  feels a bit like a magic number, should it be a variable with a comment to explain why we selected that value and why it can be greater?

Reply:
Yes, something like that?

```diff
diff --git a/pkg/controller/elasticsearch/nodespec/lifecycle_hook.go b/pkg/controller/elasticsearch/nodespec/lifecycle_hook.go
index cf91a1433..590dd6adb 100644
--- a/pkg/controller/elasticsearch/nodespec/lifecycle_hook.go
+++ b/pkg/controller/elasticsearch/nodespec/lifecycle_hook.go
@@ -83,6 +83,9 @@ function request() {
   return 0
 }

+# number of retries to try not to last more than default terminateGracePeriodSeconds (0 + 1 + 2 + 4 + 8 + 16 + 32 + 64 < 180s)
+retries_count=8
+
 function retry() {
   local retries=$1
   shift
@@ -169,7 +172,7 @@ fi
 ES_URL="{{.ServiceURL}}"

 log "retrieving node ID"
-if ! retry 8 request -X GET "${ES_URL}/_cat/nodes?full_id=true&h=id,name" "${BASIC_AUTH[@]}"
+if ! retry "$retries_count" request -X GET "${ES_URL}/_cat/nodes?full_id=true&h=id,name" "${BASIC_AUTH[@]}"
 then
   error_exit "failed to retrieve nodes"
 fi
@@ -191,7 +194,7 @@ if grep -q -v '"nodes":\[\]' "$resp_body"; then
 fi

 log "initiating node shutdown"
-if ! retry 8 request -X PUT "${ES_URL}/_nodes/${NODE_ID}/shutdown" "${BASIC_AUTH[@]}" -H 'Content-Type: application/json' -d"
+if ! retry "$retries_count" request -X PUT "${ES_URL}/_nodes/${NODE_ID}/shutdown" "${BASIC_AUTH[@]}" -H 'Content-Type: application/json' -d"
```

"try" because there are two `retry $n request` in the script. If we consumes the budget in the first, but reach the second, we can still last longer than `terminateGracePeriodSeconds`.

Reply:
In my opinion i would not make this change and make this a new pull request after this script is fixed and use one feature request at a time. I tried to not change the script functionally and only made changes for problematic scripting errors and stopped making changes after the script finally worked without errors.

In that regard it can be said that it also should return to 10 and break each change apart:

1. Fix: `error_exit` if NODE_ID is empty.
2. Fix: Style fixes and errors found by `shellcheck`
3. Fix: More sensible usage of the retry counter.

I also had a few that i didn't do:

1. Use a library for all for scripts and reuse the same code.
2. Test the code using `shellspec` and test the library every time.
3. Add more logging lines to debug.
4. Attach it to log4j and differentiate between debug and info lines.

And got more, but for now i only want NODE_ID fixed.

Reply:
I think it's better to change to `8` as you pointed out and fine to do it in this PR. I push [3285537] with the above suggestion. I'm going to rerun the tests as a matter of principle and it will be good to be merged.